File: controllers/authController.js
const bcrypt = require('bcryptjs');
const jwt = require('jsonwebtoken');
const pool = require('../config/db');
require('dotenv').config();

// Generate a 6-digit verification code
const generateCode = () => {
  return Math.floor(100000 + Math.random() * 900000).toString();
};

// Generate JWT token
const generateToken = (user) => {
  return jwt.sign(
    { id: user.id, email: user.email, role: user.role, name: user.name },
    process.env.JWT_SECRET,
    { expiresIn: process.env.JWT_EXPIRES_IN || '7d' }
  );
};

// POST /api/auth/signup
exports.signup = async (req, res) => {
  try {
    const { name, email, password, phone, role, location, businessName } = req.body;

    // Check if user already exists
    const [existing] = await pool.query('SELECT id FROM users WHERE email = ?', [email]);
    if (existing.length > 0) {
      return res.status(400).json({ error: 'User with this email already exists.' });
    }

    // Hash password
    const salt = await bcrypt.genSalt(10);
    const hashedPassword = await bcrypt.hash(password, salt);

    // Insert user
    const [result] = await pool.query(
      'INSERT INTO users (name, email, password, phone, role, location, business_name) VALUES (?, ?, ?, ?, ?, ?, ?)',
      [name, email, hashedPassword, phone || null, role, location || null, businessName || null]
    );

    // Generate verification code
    const code = generateCode();
    const expiresAt = new Date(Date.now() + 10 * 60 * 1000); // 10 minutes

    // Delete old codes for this email
    await pool.query('DELETE FROM verification_codes WHERE email = ?', [email]);

    // Insert new code
    await pool.query(
      'INSERT INTO verification_codes (email, code, expires_at) VALUES (?, ?, ?)',
      [email, code, expiresAt]
    );

    res.status(201).json({
      message: 'Account created successfully. Please verify your account.',
      userId: result.insertId,
      verificationCode: code, // In production, send via email/SMS
      email: email,
      role: role
    });
  } catch (err) {
    console.error('Signup error:', err);
    res.status(500).json({ error: 'Server error during signup.' });
  }
};

// POST /api/auth/login
exports.login = async (req, res) => {
  try {
    const { email, password, role } = req.body;

    // Find user
    const [users] = await pool.query('SELECT * FROM users WHERE email = ? AND role = ?', [email, role]);
    if (users.length === 0) {
      return res.status(400).json({ error: 'Invalid email, role, or password.' });
    }

    const user = users[0];

    // Check if user is active
    if (user.status !== 'active') {
      return res.status(403).json({ error: 'Your account has been suspended or deactivated.' });
    }

    // Verify password
    const isMatch = await bcrypt.compare(password, user.password);
    if (!isMatch) {
      return res.status(400).json({ error: 'Invalid email, role, or password.' });
    }

    // Generate verification code
    const code = generateCode();
    const expiresAt = new Date(Date.now() + 10 * 60 * 1000);

    await pool.query('DELETE FROM verification_codes WHERE email = ?', [email]);
    await pool.query(
      'INSERT INTO verification_codes (email, code, expires_at) VALUES (?, ?, ?)',
      [email, code, expiresAt]
    );

    res.json({
      message: 'Verification code sent.',
      verificationCode: code, // In production, send via email/SMS
      email: email,
      role: role,
      userName: user.name
    });
  } catch (err) {
    console.error('Login error:', err);
    res.status(500).json({ error: 'Server error during login.' });
  }
};

// POST /api/auth/verify
exports.verify = async (req, res) => {
  try {
    const { email, code } = req.body;

    // Find valid code
    const [codes] = await pool.query(
      'SELECT * FROM verification_codes WHERE email = ? AND code = ? AND used = 0 AND expires_at > NOW() ORDER BY created_at DESC LIMIT 1',
      [email, code]
    );

    if (codes.length === 0) {
      return res.status(400).json({ error: 'Invalid or expired verification code.' });
    }

    // Mark code as used
    await pool.query('UPDATE verification_codes SET used = 1 WHERE id = ?', [codes[0].id]);

    // Mark user as verified
    await pool.query('UPDATE users SET is_verified = 1 WHERE email = ?', [email]);

    // Get user data
    const [users] = await pool.query('SELECT id, name, email, role, phone, location, business_name, bio, profile_photo, status FROM users WHERE email = ?', [email]);
    const user = users[0];

    // Generate JWT
    const token = generateToken(user);

    res.json({
      message: 'Verification successful!',
      token,
      user: {
        id: user.id,
        name: user.name,
        email: user.email,
        role: user.role,
        phone: user.phone,
        location: user.location,
        businessName: user.business_name,
        bio: user.bio,
        profilePhoto: user.profile_photo,
        status: user.status
      }
    });
  } catch (err) {
    console.error('Verify error:', err);
    res.status(500).json({ error: 'Server error during verification.' });
  }
};

// POST /api/auth/resend-code
exports.resendCode = async (req, res) => {
  try {
    const { email } = req.body;

    // Check user exists
    const [users] = await pool.query('SELECT id FROM users WHERE email = ?', [email]);
    if (users.length === 0) {
      return res.status(400).json({ error: 'User not found.' });
    }

    // Generate new code
    const code = generateCode();
    const expiresAt = new Date(Date.now() + 10 * 60 * 1000);

    await pool.query('DELETE FROM verification_codes WHERE email = ?', [email]);
    await pool.query(
      'INSERT INTO verification_codes (email, code, expires_at) VALUES (?, ?, ?)',
      [email, code, expiresAt]
    );

    res.json({
      message: 'New verification code generated.',
      verificationCode: code
    });
  } catch (err) {
    console.error('Resend code error:', err);
    res.status(500).json({ error: 'Server error.' });
  }
};

// GET /api/auth/me
exports.getProfile = async (req, res) => {
  try {
    const [users] = await pool.query(
      'SELECT id, name, email, role, phone, location, business_name, bio, profile_photo, status, created_at FROM users WHERE id = ?',
      [req.user.id]
    );

    if (users.length === 0) {
      return res.status(404).json({ error: 'User not found.' });
    }

    const user = users[0];
    res.json({
      id: user.id,
      name: user.name,
      email: user.email,
      role: user.role,
      phone: user.phone,
      location: user.location,
      businessName: user.business_name,
      bio: user.bio,
      profilePhoto: user.profile_photo,
      status: user.status,
      createdAt: user.created_at
    });
  } catch (err) {
    console.error('Get profile error:', err);
    res.status(500).json({ error: 'Server error.' });
  }
};

// PUT /api/auth/profile
exports.updateProfile = async (req, res) => {
  try {
    const { name, phone, location, businessName, bio } = req.body;

    await pool.query(
      'UPDATE users SET name = ?, phone = ?, location = ?, business_name = ?, bio = ? WHERE id = ?',
      [name, phone || null, location || null, businessName || null, bio || null, req.user.id]
    );

    // Return updated user
    const [users] = await pool.query(
      'SELECT id, name, email, role, phone, location, business_name, bio, profile_photo, status FROM users WHERE id = ?',
      [req.user.id]
    );
    const user = users[0];

    res.json({
      message: 'Profile updated successfully.',
      user: {
        id: user.id,
        name: user.name,
        email: user.email,
        role: user.role,
        phone: user.phone,
        location: user.location,
        businessName: user.business_name,
        bio: user.bio,
        profilePhoto: user.profile_photo,
        status: user.status
      }
    });
  } catch (err) {
    console.error('Update profile error:', err);
    res.status(500).json({ error: 'Server error.' });
  }
};

File: backend-spring/src/main/java/com/agrivalue/controller/AuthController.java
package com.agrivalue.controller;

import com.agrivalue.entity.User;
import com.agrivalue.entity.VerificationCode;
import com.agrivalue.repository.UserRepository;
import com.agrivalue.repository.VerificationCodeRepository;
import com.agrivalue.security.JwtUserPrincipal;
import com.agrivalue.security.JwtUtil;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.*;

@RestController
@RequestMapping("/api/auth")
public class AuthController {

    private final UserRepository userRepo;
    private final VerificationCodeRepository codeRepo;
    private final PasswordEncoder passwordEncoder;
    private final JwtUtil jwtUtil;

    public AuthController(UserRepository userRepo, VerificationCodeRepository codeRepo,
                          PasswordEncoder passwordEncoder, JwtUtil jwtUtil) {
        this.userRepo = userRepo;
        this.codeRepo = codeRepo;
        this.passwordEncoder = passwordEncoder;
        this.jwtUtil = jwtUtil;
    }

    private String generateCode() {
        return String.valueOf(100000 + new Random().nextInt(900000));
    }

    @PostMapping("/signup")
    @Transactional
    public ResponseEntity<?> signup(@RequestBody Map<String, String> body) {
        String email = body.get("email");
        if (userRepo.existsByEmail(email)) {
            return ResponseEntity.badRequest().body(Map.of("error", "User with this email already exists."));
        }

        User user = User.builder()
                .name(body.get("name"))
                .email(email)
                .password(passwordEncoder.encode(body.get("password")))
                .phone(body.get("phone"))
                .role(User.Role.valueOf(body.get("role")))
                .location(body.get("location"))
                .businessName(body.get("businessName"))
                .build();
        user = userRepo.save(user);

        String code = generateCode();
        codeRepo.deleteByEmail(email);
        codeRepo.save(VerificationCode.builder()
                .email(email).code(code)
                .expiresAt(LocalDateTime.now().plusMinutes(10))
                .build());

        return ResponseEntity.status(HttpStatus.CREATED).body(Map.of(
                "message", "Account created successfully. Please verify your account.",
                "userId", user.getId(),
                "verificationCode", code,
                "email", email,
                "role", body.get("role")
        ));
    }

    @PostMapping("/login")
    @Transactional
    public ResponseEntity<?> login(@RequestBody Map<String, String> body) {
        String email = body.get("email");
        String role = body.get("role");

        Optional<User> optUser = userRepo.findByEmailAndRole(email, User.Role.valueOf(role));
        if (optUser.isEmpty()) {
            return ResponseEntity.badRequest().body(Map.of("error", "Invalid email, role, or password."));
        }

        User user = optUser.get();
        if (user.getStatus() != User.Status.active) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).body(Map.of("error", "Your account has been suspended or deactivated."));
        }
        if (!passwordEncoder.matches(body.get("password"), user.getPassword())) {
            return ResponseEntity.badRequest().body(Map.of("error", "Invalid email, role, or password."));
        }

        String code = generateCode();
        codeRepo.deleteByEmail(email);
        codeRepo.save(VerificationCode.builder()
                .email(email).code(code)
                .expiresAt(LocalDateTime.now().plusMinutes(10))
                .build());

        return ResponseEntity.ok(Map.of(
                "message", "Verification code sent.",
                "verificationCode", code,
                "email", email,
                "role", role,
                "userName", user.getName()
        ));
    }

    @PostMapping("/verify")
    @Transactional
    public ResponseEntity<?> verify(@RequestBody Map<String, String> body) {
        String email = body.get("email");
        String code = body.get("code");

        Optional<VerificationCode> optCode = codeRepo
                .findFirstByEmailAndCodeAndUsedFalseAndExpiresAtAfterOrderByCreatedAtDesc(email, code, LocalDateTime.now());
        if (optCode.isEmpty()) {
            return ResponseEntity.badRequest().body(Map.of("error", "Invalid or expired verification code."));
        }

        VerificationCode vc = optCode.get();
        vc.setUsed(true);
        codeRepo.save(vc);

        Optional<User> optUser = userRepo.findByEmail(email);
        if (optUser.isEmpty()) {
            return ResponseEntity.badRequest().body(Map.of("error", "User not found."));
        }

        User user = optUser.get();
        user.setIsVerified(true);
        userRepo.save(user);

        String token = jwtUtil.generateToken(user.getId(), user.getEmail(), user.getRole().name(), user.getName());

        Map<String, Object> userData = new LinkedHashMap<>();
        userData.put("id", user.getId());
        userData.put("name", user.getName());
        userData.put("email", user.getEmail());
        userData.put("role", user.getRole().name());
        userData.put("phone", user.getPhone());
        userData.put("location", user.getLocation());
        userData.put("businessName", user.getBusinessName());
        userData.put("bio", user.getBio());
        userData.put("profilePhoto", user.getProfilePhoto());
        userData.put("status", user.getStatus().name());

        return ResponseEntity.ok(Map.of("message", "Verification successful!", "token", token, "user", userData));
    }

    @PostMapping("/resend-code")
    @Transactional
    public ResponseEntity<?> resendCode(@RequestBody Map<String, String> body) {
        String email = body.get("email");
        if (!userRepo.existsByEmail(email)) {
            return ResponseEntity.badRequest().body(Map.of("error", "User not found."));
        }

        String code = generateCode();
        codeRepo.deleteByEmail(email);
        codeRepo.save(VerificationCode.builder()
                .email(email).code(code)
                .expiresAt(LocalDateTime.now().plusMinutes(10))
                .build());

        return ResponseEntity.ok(Map.of("message", "New verification code generated.", "verificationCode", code));
    }

    @GetMapping("/me")
    public ResponseEntity<?> getProfile(@AuthenticationPrincipal JwtUserPrincipal principal) {
        Optional<User> optUser = userRepo.findById(principal.getId());
        if (optUser.isEmpty()) return ResponseEntity.notFound().build();

        User user = optUser.get();
        Map<String, Object> data = new LinkedHashMap<>();
        data.put("id", user.getId());
        data.put("name", user.getName());
        data.put("email", user.getEmail());
        data.put("role", user.getRole().name());
        data.put("phone", user.getPhone());
        data.put("location", user.getLocation());
        data.put("businessName", user.getBusinessName());
        data.put("bio", user.getBio());
        data.put("profilePhoto", user.getProfilePhoto());
        data.put("status", user.getStatus().name());
        data.put("createdAt", user.getCreatedAt());
        return ResponseEntity.ok(data);
    }

    @PutMapping("/profile")
    public ResponseEntity<?> updateProfile(@AuthenticationPrincipal JwtUserPrincipal principal,
                                           @RequestBody Map<String, String> body) {
        Optional<User> optUser = userRepo.findById(principal.getId());
        if (optUser.isEmpty()) return ResponseEntity.notFound().build();

        User user = optUser.get();
        user.setName(body.getOrDefault("name", user.getName()));
        user.setPhone(body.get("phone"));
        user.setLocation(body.get("location"));
        user.setBusinessName(body.get("businessName"));
        user.setBio(body.get("bio"));
        userRepo.save(user);

        Map<String, Object> userData = new LinkedHashMap<>();
        userData.put("id", user.getId());
        userData.put("name", user.getName());
        userData.put("email", user.getEmail());
        userData.put("role", user.getRole().name());
        userData.put("phone", user.getPhone());
        userData.put("location", user.getLocation());
        userData.put("businessName", user.getBusinessName());
        userData.put("bio", user.getBio());
        userData.put("profilePhoto", user.getProfilePhoto());
        userData.put("status", user.getStatus().name());

        return ResponseEntity.ok(Map.of("message", "Profile updated successfully.", "user", userData));
    }
}
